alternator/streams.cc

```diff
@@ -220,38 +222,59 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
 
 struct shard_id {
     static constexpr auto marker = 'H';
-
-    utils::UUID table;
-    db_clock::time_point time;
-    cdc::stream_id id;
-
-    shard_id(utils::UUID tab, db_clock::time_point t, cdc::stream_id i)
-        : table(tab)
-        , time(t)
-        , id(i)
-    {}
+    static constexpr auto hex_width = sizeof(uint64_t)*8/4;
+
+    db_clock::time_point start_time;
+    db_clock::time_point end_time; // zero if open
+    dht::token lo_token; // range start, inclusive
+    dht::token hi_token; // range end, exclusive
+
+    shard_id(db_clock::time_point start,  db_clock::time_point end, dht::token lo, dht::token hi)
+        : start_time(start)
+        , end_time(end)
+        , lo_token(lo)
+        , hi_token(hi)
+    {
+        assert(lo_token < hi_token);
+        assert(end_time == db_clock::time_point{} || start_time < end_time);
+    }
     shard_id(const sstring&);
 
+    bool has_expired() const {
+        return end_time != db_clock::time_point{};
+    }
+
+    // dynamo specifies shardid as max 65 chars. Thus we cannot use separators, 
+    // instead we have to rely on 16 width hex
     friend std::ostream& operator<<(std::ostream& os, const shard_id& id) {
         boost::io::ios_flags_saver fs(os);
-        return os << marker << id.table << std::hex
-            << ':' << id.time.time_since_epoch().count()
-            << ':' << id.id.to_bytes()
+        return os << marker << std::hex << std::setfill('0') 
+            << std::setw(hex_width) << id.start_time.time_since_epoch().count()
+            << std::setw(hex_width) << id.end_time.time_since_epoch().count()
+            << std::setw(hex_width) << (id.lo_token.is_minimum() ? 0ull : dht::token::to_int64(id.lo_token))
+            << std::setw(hex_width) << (id.hi_token.is_maximum() ? 0ull : dht::token::to_int64(id.hi_token))
             ;
     }
 };
 
 shard_id::shard_id(const sstring& s) {
-    auto i = s.find(':');
-    auto j = s.find(':', i + 1);
-
-    if (s.at(0) != marker || i == sstring::npos || j == sstring::npos) {
+    if (s.at(0) != marker || s.size() != (1 + hex_width * 4)) {
         throw std::invalid_argument(s);
     }
 
-    table = utils::UUID(s.substr(1, i));
-    time = db_clock::time_point(db_clock::duration(std::stoull(s.substr(i + 1, j), nullptr, 16)));
-    id = cdc::stream_id(from_hex(s.substr(j + 1)));
+    start_time = db_clock::time_point(db_clock::duration(std::stoull(s.substr(1, hex_width), nullptr, 16)));
+    end_time = db_clock::time_point(db_clock::duration(std::stoull(s.substr(1 + hex_width, hex_width), nullptr, 16)));
+    auto lo = std::stoull(s.substr(1 + 2*hex_width, hex_width), nullptr, 16);
+    auto hi = std::stoull(s.substr(1 + 3*hex_width), nullptr, 16);
+    lo_token = lo == 0 ? dht::minimum_token() : dht::token::from_int64(lo);
+    hi_token = hi == 0 ? dht::maximum_token() : dht::token::from_int64(hi);
+
+    if (end_time != db_clock::time_point{} && end_time < start_time) {
```

Comment:
Super nit: s/</<=/

Reply:
No, I purposely allow for end to be equal to start. While our cdc generation will not do this, it is a valid state for the ID here, so it is acceptable. It will not yield data, but acceptable.